src/main/java/org/truffleruby/core/array/ArrayWriteNormalizedNode.java

```diff
@@ -88,11 +135,25 @@ public Object writeBeyondPrimitive(DynamicObject array, int index, Object value,
     }
 
     @Specialization(guards = {
-            "!isInBounds(array, index)", "!isExtendingByOne(array, index)", "strategy.matches(array)", "!strategy.isPrimitive()"
+            "strategy.isStorageMutable()", "!isInBounds(array, index)", "!isExtendingByOne(array, index)", "strategy.matches(array)", "!strategy.isPrimitive()"
     }, limit = "STORAGE_STRATEGIES")
     public Object writeBeyondObject(DynamicObject array, int index, Object value,
             @Cached("of(array)") ArrayStrategy strategy,
             @Cached("create()") ArrayEnsureCapacityNode ensureCapacityNode) {
+        return writeBeyondObjectInternal(array, index, value, strategy, ensureCapacityNode);
+    }
+
+    @Specialization(guards = {
+            "!strategy.isStorageMutable()", "!isInBounds(array, index)", "!isExtendingByOne(array, index)", "strategy.matches(array)", "!mutableStrategy.isPrimitive()"
+    }, limit = "STORAGE_STRATEGIES")
+    public Object writeBeyondObject(DynamicObject array, int index, Object value,
+            @Cached("of(array)") ArrayStrategy strategy,
+            @Cached("strategy.generalizeForMutation()") ArrayStrategy mutableStrategy,
+            @Cached("create()") ArrayEnsureCapacityNode ensureCapacityNode) {
+        return writeBeyondObjectInternal(array, index, value, mutableStrategy, ensureCapacityNode);
+    }
```

Comment:
Can this be merged with the one above by always using `generalizeForMutation()`?